### PR TITLE
Add support for external libraries to extracted semantics tests.

### PR DESCRIPTION
**Pull request by ekpyron**
Prerequisite for properly testing https://github.com/ethereum/solidity/issues/6420 in extracted tests.

We can change the syntax later, if anybody is unhappy with it...

### CODE REVIEW

File: test/libsolidity/util/TestFileParser.cpp [hunk 1]
```diff
@@ -75,43 +75,55 @@ vector<dev::solidity::test::FunctionCall> TestFileParser::parseFunctionCalls(siz
 
 				try
 				{
-					tie(call.signature, call.useCallWithoutSignature) = parseFunctionSignature();
```
Comment (ekpyron) [inline test/libsolidity/util/TestFileParser.cpp:78]: This is mostly whitespace changes, so with ``Hide whitespace changes`` set, the change is rather trivial.

File: test/libsolidity/SemanticTest.h [hunk 2]
```diff
@@ -60,7 +60,7 @@ class SemanticTest: public SolidityExecutionFramework, public EVMVersionRestrict
 
 	/// Compiles and deploys currently held source.
 	/// Returns true if deployment was successful, false otherwise.
-	bool deploy(std::string const& _contractName, u256 const& _value, bytes const& _arguments);
+	bool deploy(std::string const& _contractName, u256 const& _value, bytes const& _arguments, std::map<std::string, dev::test::Address> const &_libraries = {});
```
Comment (leonardoalt) [inline test/libsolidity/SemanticTest.h]:
```suggestion
	bool deploy(std::string const& _contractName, u256 const& _value, bytes const& _arguments, std::map<std::string, dev::test::Address> const& _libraries = {});
```

File: test/libsolidity/SemanticTest.cpp [hunk 3]
```diff
@@ -78,15 +78,35 @@ TestCase::TestResult SemanticTest::run(ostream& _stream, string const& _linePref
...
+						"Failed to deploy library " + test.call().signature
+					);
+					libraries[test.call().signature] = m_contractAddress;
+					continue;
+				}
 				else
-					soltestAssert(deploy("", 0, bytes()), "Failed to deploy contract.");
+				{
+					if (test.call().isConstructor)
+						deploy("", test.call().value, test.call().arguments.rawBytes(), libraries);
```
Comment (leonardoalt) [inline test/libsolidity/SemanticTest.cpp]: Maybe not for this PR, but one could also sort a copy of `m_tests` before such that libraries would be deployed first

File: test/libsolidity/SemanticTest.cpp [hunk 4]
```diff
@@ -78,15 +78,35 @@ TestCase::TestResult SemanticTest::run(ostream& _stream, string const& _linePref
...
+		map<string, dev::test::Address> libraries;
+
+		bool constructed = false;
+
 		for (auto& test: m_tests)
 		{
-			if (&test == &m_tests.front())
-				if (test.call().isConstructor)
-					deploy("", test.call().value, test.call().arguments.rawBytes());
+			if (!constructed)
```
Comment (leonardoalt) [inline test/libsolidity/SemanticTest.cpp]: I think the coding style says that if you use `if-else` the `if` condition should not be negated.

File: test/libsolidity/semanticTests/libraries/stub.sol [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+contract C {
+    function g(uint256 v) external returns (uint256) {
+        return L.f(v);
+    }
+}
+// ----
+// library: L
+// g(uint256): 1 -> 1
+// g(uint256): 2 -> 4
+// g(uint256): 4 -> 16
```
Comment (leonardoalt) [inline test/libsolidity/semanticTests/libraries/stub.sol:13]: Could it also call `f` directly?

File: test/libsolidity/semanticTests/libraries/stub.sol [hunk 6]
```diff
@@ -0,0 +1,13 @@
+library L {
+    function f(uint256 v) external returns (uint256) { return v*v; }
+}
+contract C {
+    function g(uint256 v) external returns (uint256) {
+        return L.f(v);
+    }
+}
+// ----
+// library: L
```
Comment (leonardoalt) [inline test/libsolidity/semanticTests/libraries/stub.sol:10]: What exactly is the syntax representing? That the lib used in `g` is `L`?

File: test/libsolidity/SemanticTest.cpp (hunk 3 above)
Comment (ekpyron) [reply test/libsolidity/SemanticTest.cpp]: Yeah - or parse libraries and constructors separately from the actual calls... but I'd postpone that.

File: test/libsolidity/semanticTests/libraries/stub.sol (hunk 5 above)
Comment (ekpyron) [reply test/libsolidity/semanticTests/libraries/stub.sol:13]: Nope, I don't think so.

File: test/libsolidity/semanticTests/libraries/stub.sol (hunk 6 above)
Comment (ekpyron) [reply test/libsolidity/semanticTests/libraries/stub.sol:10]: This says "Deploy a copy of library L and link all contracts against it, i.e. set the address of the link references to L to the address of that deployed copy of L"...

Comment (ekpyron) [reply test/libsolidity/semanticTests/libraries/stub.sol:10]: I thought about looking for all libraries in the source and deploying and resolving them automatically, but then we could for example not test that using only internal library functions doesn't require linking, etc...

File: test/libsolidity/semanticTests/libraries/stub.sol (hunk 5 above)
Comment (ekpyron) [reply test/libsolidity/semanticTests/libraries/stub.sol:13]: As in - the calls always go to the last contract in the source, which is C not L.

File: test/libsolidity/SemanticTest.cpp (hunk 4 above)
Comment (ekpyron) [reply test/libsolidity/SemanticTest.cpp]: Does it? Ok, fine with me :-).

File: test/libsolidity/semanticTests/libraries/stub.sol (hunk 6 above)
Comment (leonardoalt) [reply test/libsolidity/semanticTests/libraries/stub.sol:10]: Ah, right. Can you add the same test but without this line?

Comment (ekpyron) [reply test/libsolidity/semanticTests/libraries/stub.sol:10]: In fact I had that earlier, but ``soltest`` does have a check in place ensuring that the link references are empty when deploying the contract, so not having this line will fail an assertion....

Comment (ekpyron) [reply test/libsolidity/semanticTests/libraries/stub.sol:10]: This one: https://github.com/ethereum/solidity/blob/50a015acf0c270d2cf9070d9e9d9178bd38a0dca/test/libsolidity/SolidityExecutionFramework.cpp#L86

Comment (ekpyron) [reply test/libsolidity/semanticTests/libraries/stub.sol:10]: And at the moment we can't gracefully check for that error being thrown... would probably be quite hacky at the moment and involve catching that exception here and treating it specially... the problem is that ``SolidityExecutionFramework`` is still shared between the new extracted SemanticsTests and the old SolidityEndToEndTests.cpp ones... and for the latter I'd rather not remove this assertion... do you think it's worth spending some effort on this for adding that test?

Comment (leonardoalt) [reply test/libsolidity/semanticTests/libraries/stub.sol:10]: Ah ok, fine without. I guess we have tests for internal library calls without linking?

**Review by leonardoalt — DISMISSED** [review]

File: test/libsolidity/semanticTests/libraries/stub.sol (hunk 6 above)
Comment (ekpyron) [reply test/libsolidity/semanticTests/libraries/stub.sol:10]: Hm... we might actually not - no extracted ones for sure, but I don't see any non-extracted ones either... I'll add one here.

Comment (ekpyron) [reply test/libsolidity/semanticTests/libraries/stub.sol:10]: Added.

**Comment by ekpyron** [comment]
Added internal library call test without linking.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**